### ISSUE TITLE
[Bug]: When I upgraded to version `0.12.x` I couldn't get the markdown `formatter`

### ISSUE THREAD

**Issue by XeryYue** [antfu/vite-plugin-md#64]
# Description

In ver `0.11.7`

When i use dynamic import load `.md` formatter info. i can got them.

![image](https://user-images.githubusercontent.com/52351095/162939665-075fb85a-db32-4b34-97d9-dded516b8ff4.png)

This is my chrome console

![image](https://user-images.githubusercontent.com/52351095/162939968-fa6e354c-235b-4153-991e-5819bdf08c09.png)

After ver `0.11.7` in new version `0.12.4`

In my console

![image](https://user-images.githubusercontent.com/52351095/162940480-e2ee2d82-eb87-4148-8b3d-1810644cba81.png)

@antfu  PTAL 🙏

**Comment by yankeeinlondon** [comment]
thanks for bringing this to my attention ... i've a unit test now which shows this failing (I don't personally use this atm) and will see what I can do to get this fixed soon.

**Comment by yankeeinlondon** [comment]
@antfu the immediate error is a bit hard to grok and I wouldn't have thought that any of the changes in version `0.12.x` would have had any direct impact on this but then my debugging skills inside the Vite build pipeline is something I need to get better at. Any suggestions on how best to debug?

![image](https://user-images.githubusercontent.com/980757/164566362-72d2608a-0bd0-48bf-82c9-66ed43b6ef55.png)

**Comment by yankeeinlondon** [comment]
I just tried running this test using `0.11.9`:

- test file:
  ```ts
  import { describe, expect, it } from 'vitest'
  import MarkdownComponent, { frontmatter, title } from './test/fixtures/using-frontmatter.md'

  describe('exposeFrontmatter property', () => {
    it('markdown files offer frontmatter properties for import', async() => {
      expect(MarkdownComponent).toBeDefined()
      expect(frontmatter).toBeDefined()
      expect(title).toBeDefined()
    })
  })
  ```
- Added a shim file
  ```ts
  declare module '*.md' {
    import { Component } from 'vue'

    const Component: Component
    export default Component
    export const frontmatter: Record<string, unknown>
    export const title: string | undefined
    export const description: string | undefined
    export const layout: string | undefined
  }
  ```
- I then added a `vitest.config.ts` like we have in the `0.12.x` releases (and installed pages and layouts as devdeps):
  ```ts
  /// <reference types="vitest" />
  import { defineConfig } from 'vite'
  import Pages from 'vite-plugin-pages'
  import Layouts from 'vite-plugin-vue-layouts'
  import Markdown from './src/index'

  // used for testing, library code uses TSUP to build exports
  export default defineConfig(() => ({
    test: {
      dir: 'test',
    },
    plugins: [
      Pages({
        extensions: ['vue', 'md'],
      }),
      Layouts(),
      Markdown(),
    ],
  }))
  ```
- At this point I was expecting that it would work but instead I get this error:
  ![image](https://user-images.githubusercontent.com/980757/164571254-d1ef1c72-c316-4b09-939b-0dabb0b7d7a2.png)

- I will now go back to `0.11.7` as that's the version @XeryYue was referencing

**Comment by yankeeinlondon** [comment]
same error on version `0.11.7` ... maybe I'm going down the wrong path

**Comment by yankeeinlondon** [comment]
Since the errors seem to express some fault with the generated VueJS component's structure I thought it would be worth looking at this from the `0.12.x` series (in this case Vue3):

```vue
<template>
  <div class="markdown-body">
    <h1>{{ title }}</h1>
    <p>A nice image to brighten your day:</p>
    <p><img src="%7B%7Bimage%7D%7D" alt="test image"></p>
  </div>
</template>
<script setup>
const frontmatter = { "image": "/images/foobar.jpg", "title": "Frontmatter in Action" }
const image = "/images/foobar.jpg"
const title = "Frontmatter in Action"
const excerpt = undefined
defineExpose({ frontmatter })
const head = { "image": "/images/foobar.jpg", "title": "Frontmatter in Action" }
</script>
```

This looks perfectly valid. Which makes me think that possibly it's the upgraded versions of **Vite**?

- Vite: `2.7.6` -> `2.9.5`

That might just be being optimistic but do you see any issue with the markdown files. The snapshot tests are more complete now so there's no exact "apples for apples" but  honestly the `0.12.x` ones just seem more optimized but largely the same.

**Comment by XeryYue** [comment]
> thanks for bringing this to my attention ... i've a unit test now which shows this failing (I don't personally use this atm) and will see what I can do to get this fixed soon.

thanks, :)

**Comment by yankeeinlondon** [comment]
if you saw some other comments ... I appologize as I was thinking that your issue was related to a different issue regarding importing **frontmatter** properties. They may indeed be related but I think I'm going to need more information from you.

What would be best is if you can create a minimally reproducible example of this on [StackBlitz playground](https://vite.new/vue-ts). Also when you say "formatter info" what are you referring to? Your example seems to revolve around frontmatter metadata and I do wonder if there was an undocumented change in version `0.12.x` (in fact it was probably `0.11.8`) which caused this.

If I'm right in my theory then there has been no loss in _utility_ but there was change in a semi-formal part of the API contract.  Let me illustrate with an example:
- if we define `title` and `meta` properties these are always exposed on the `frontmatter` object
- we expose them by using VueJS's `defineExpose( { frontmatter} )`
- in the component's script code we also expose the variables as bare properties that the given can page can use directly without needing to always reference "frontmatter"
- these bare properties, however, are not put into the `defineExpose()` function and would duplicative if they were
- that said, I have the vague memory of maybe this was being done in the past but it's likely not a good idea and the tests from that era didn't expose this "feature" in the very limited tests that were in place then

**Comment by yankeeinlondon** [comment]
here's an example of what the markdown files from version `0.12.x` are converted to:

```vue
<template><div class="markdown-body"><h1>{{title}}</h1>
<p>A nice image to brighten your day:</p>
<p><img src="%7B%7Bimage%7D%7D" alt="test image"></p>
</div>
</template>
<script setup>const frontmatter = {"image":"/images/foobar.jpg","title":"Frontmatter in Action"}
 const image = "/images/foobar.jpg"
 const title = "Frontmatter in Action"
 const excerpt = undefined
defineExpose({ frontmatter })
const head = {"image":"/images/foobar.jpg","title":"Frontmatter in Action"}
</script>

**Comment by yankeeinlondon** [comment]
Does what I'm describing make sense and does it map to the issue you're having?

**Comment by XeryYue** [comment]
> Does what I'm describing make sense and does it map to the issue you're having?

yeah, And i will take a minimally exmaple :)

**Comment by XeryYue** [comment]
Hi, Here is a `0.11.x` [exmaple](https://stackblitz.com/edit/vitejs-vite-lgeh8f?file=package-lock.json)

**Comment by yankeeinlondon** [comment]
fantasic reproduction, thanks @XeryYue . WIll look over now and hopefully get this turned around. It dips into functionality I had definitely not built tests around and not personally used but excited to get this sorted and get some tests around it so it won't regress in the future.

**Comment by yankeeinlondon** [comment]
@XeryYue @antfu might need your collective brain power on how best to create a good unit test for this. I figured that this test:

```ts
describe('validate that imports of frontmatter are possible', () => {
  it('title is found', async() => {
    const md = readFileSync('test/fixtures/my-markdown.md', 'utf-8')
    const comp = markdownToVue('my-markdown.md', md)
    const imp = (await import('test/fixtures/my-markdown.md')) as { default: Component; title: string }
    expect(imp.default).toBeDefined()
    expect(imp.title).toBeDefined()
  })
})
```

along with this shim file:

```ts
declare module '*.md' {
  import { Component } from 'vue'
  const Component: Component
  export const frontmatter: Record<string, unknown>
  export const title: string
  export default Component
}
````

Should be just fine but when I do this using `0.11.7` I still get this error:

> Error: Failed to parse source for import analysis because the content contains invalid JS syntax. You may need to install appropriate plugins to handle the .md file format.

Which at first I presumed to be because the library actually uses TSUP to build not ViteJS and so I presume Vitest wasn't able to reach this plugin's support for MD files. I thought i'd corrected this by putting the the following vite config file in the root of the directory but I still get the same error:

```ts
/// <reference types="vitest" />
import { defineConfig } from 'vite'
import Markdown from './src/index'

// used for testing, library code uses TSUP to build exports
export default defineConfig(() => ({
  test: {
    dir: 'test',
  },
  plugins: [
    Markdown(),
  ],
}))
```

**Comment by yankeeinlondon** [comment]
I believe I should have a fix for this soon but still keen on getting a more direct test for this so if either of you have any ideas how to test the import from **node** land I'd appreciate your help.

**Comment by yankeeinlondon** [comment]
I have submitted a fix which should make this work.

Note, however, the original implementation was not done quite to the specification it suggested in docs. Meaning ... the frontmatter props were exported as root level named exports on the module. This is undesirable because it can cause namespace collisions and because it makes typing the bundle more troublesome with shims. It also is inconsistent with the way `defineExpose()` presents these properties. 

> [PR](https://github.com/antfu/vite-plugin-md/pull/67)

![image](https://user-images.githubusercontent.com/980757/164836387-d03b08cb-4ab2-4847-8601-0e9475f1fc21.png)


A markdown file now is converted to a structure like this:

```vue
<script setup>
  defineExpose({ frontmatter: {"title":"Hello World","description":"testing is the path to true happiness"} })
  
 const title = "Hello World"
 const description = "testing is the path to true happiness"
 const excerpt = undefined
</script>
    
<script>
  export const frontmatter = {"title":"Hello World","description":"testing is the path to true happiness"}
</script>
```

**Comment by yankeeinlondon** [comment]
ok @antfu has released it now as `v0.13.0`; let me know if it fixes your issues. On my modification of your example code it appears to work: https://stackblitz.com/edit/vitejs-vite-lgeh8d-n1sz3c?file=package.json

@antfu I'm still concerned that we can't import the markdown files in NodeJS (that's true in `11.x` as well as `12/13.x`) and to make better unit tests I'd need to do this. I remember reading that **vitest** can operate in Vite and non-Vite environements. I added a `vite.config.ts` file to get vitest to run with ViteJS but maybe this isn't happening? We still get this error:
![image](https://user-images.githubusercontent.com/980757/164943969-13de0c6a-b75b-4675-a67d-84f9d8cd0a9c.png)

**Comment by XeryYue** [comment]
@yankeeinlondon  It's back to work again. Thank you very much : )
